src/core/qgsrendercontext.h

```diff
@@ -1108,6 +1125,9 @@ class CORE_EXPORT QgsRenderContext : public QgsTemporalRangeObject
 
     Qgis::RendererUsage mRendererUsage = Qgis::RendererUsage::Unknown;
 
+    //! clip paths to be applied to the symbol layer before rendering
+    QMap< const QgsSymbolLayer *, QList<const QPainterPath *> > mSymbolLayerClipPaths;
```

Comment:
What's happening with ownership here? 

QPainterPath is implicitly shared, so you could remove all the raw pointer use of these and safely just pass around QPainterPath values instead, with no added cost (and a lot more memory safety!)

Reply:
I cannot use QPainterPath values, because they're gonna me modified by QgsMaskPaintDevice through the mask painter during the first pass jobs. But when I [set them](https://github.com/troopa81/QGIS/blob/selective_masking_vectorization/src/core/maprenderer/qgsmaprendererjob.cpp#L887), they are not modified yet.

QPainterPath will be destroyed by QgsMaskPaintDevice destructor. However, there was a memory leak because maskPainter device was not properly deleted when force vector output was on. I fixed it [94fe61f](https://github.com/qgis/QGIS/pull/47459/commits/94fe61f6c8630d17a4f8d9a1643fe7751f5c13f8)

Reply:
See related note in qgsmaprendererjob.cpp